.vscode/extensions.json

```diff
@@ -1,3 +0,0 @@
-{
-  "recommendations": ["orta.vscode-twoslash-queries"]
-}
```

Comment:
please restore this file.  It's simply an extension recommendation that's used in the [examples](https://github.com/bufbuild/connect-query/blob/1bfc49a8d9e7232351c1c23105cbb9042d2443b1/examples/react/basic-without-generation/src/example.tsx#L29) to demonstrate the flowing of the types (which is an important part of the types).  [It](https://www.typescriptlang.org/dev/twoslash/)'s a quite common thing (used extensively in many TypeScript repos (including for testing) and the TypeScript playground.

These examples are written to be portable to something like CodeSandbox as well as the TypeScript playground (which has twoslash out-of-the-box) where this will be the end-user experience.

Reply:
I'd like to see some examples of repos that do this, our current OSS guidelines say to exclude specific IDE files in any repo, as when many people come to OSS repos and see them with specific IDE files in them, it is seen as a code smell. connect-web and protobuf-es do not do this either, so would also like to verify with @timostamm. Not saying we can't have it if it really makes sense, but we'd need to update our guidelines for TypeScript as well

Reply:
I see IDE files as code smell too. What's the practical benefit of this specific file? I'm not familiar with twoslash queries.

Reply:
> I'd like to see some examples of repos that do this

- [next.js](https://github.com/vercel/next.js/tree/canary/.vscode) 99.1k stars
- [TypeScript](https://github.com/microsoft/TypeScript/tree/main/.vscode) 87.5k stars
- [MUI](https://github.com/mui/material-ui/tree/master/.vscode) 83.9k stars
- [Puppeteer](https://github.com/puppeteer/puppeteer/tree/main/.vscode) 81.5k stars
- [storybook.js](https://github.com/storybookjs/storybook/tree/next/.vscode) 76k stars
- [tauri](https://github.com/tauri-apps/tauri/tree/dev/.vscode) 57.1k stars
- [HomeAssistant](https://github.com/home-assistant/core/tree/dev/.vscode) 57.3k stars
- [mermaid](https://github.com/mermaid-js/mermaid/tree/develop/.vscode) 53k stars
- [Parcel.js](https://github.com/parcel-bundler/parcel/tree/v2/.vscode) 41.8k stars
- [babel.js](https://github.com/babel/babel/tree/main/.vscode) 41.8k stars
- [RocketChat](https://github.com/RocketChat/Rocket.Chat/tree/develop/.vscode) 34.5k stars
- [bun](https://github.com/oven-sh/bun/tree/main/.vscode) 37.8k stars
- [astro](https://github.com/withastro/astro/tree/main/.vscode) 24.6k stars
- [Babylon.js](https://github.com/BabylonJS/Babylon.js/tree/master/.vscode) 19.2k
- [tRPC](https://github.com/trpc/trpc/tree/main/.vscode) 19.1k stars
- [expo](https://github.com/expo/expo/tree/main/.vscode) 18.7k stars
- [tensorflow.js](https://github.com/tensorflow/tfjs/tree/master/.vscode) 16.9k stars
- [fluentui](https://github.com/microsoft/fluentui/tree/master/.vscode) 14.6k stars
- [create-t3-app](https://github.com/t3-oss/create-t3-app/tree/next/.vscode) 13.1k stars
- [typescript-eslint](https://github.com/typescript-eslint/typescript-eslint/tree/main/.vscode) 12.8k stars
- [codeql](https://github.com/github/codeql/tree/main/.vscode) 5.6k stars

Reply:
We're re-adding it because we think this is minor, but issues of consistency and code smell remain, so we may want to revisit this in the future.

Reply:
Dimitri, being unfamiliar with twoslash-queries, and not a user of vscode, I wonder what I need to do if I want to contribute to this repository. Can we have an explanation here that helps?

Reply:
> being unfamiliar with twoslash-queries
> - @timostamm

Twoslash notation is used in all of the TypeScript documentation (e.g. look for the gray boxes on [this](https://www.typescriptlang.org/docs/handbook/2/conditional-types.html) page and click "Try Me" in the bottom right if you hover over the code block).  I gave a link in my first message above regarding this project's usage of it.  As I said above, even without this extension, any future place this code lands (such as CodeSandbox) will have twoslash enabled out-of-the-box anyway.

> and not a user of vscode
> - @timostamm

This specific extension recommendation is not absolutely mission critical to anything today, but I expect that since perhaps as much as [86% of JavaScript developers use vscode](https://2020.stateofjs.com/en-US/other-tools/#text_editors), it makes sense to meet our users where they are and provide launch configurations and other things they can use to get started quickly in the way that's best for them.

> I wonder what I need to do if I want to contribute to this repository
> - @timostamm

I'm sorry, I don't understand how this question connects to the rest of the conversation.  I've made a ticket to add a CONTRIBUTING.md.  Please let me know if that's not what you were asking for.

Reply:
I get that Twoslash has some decent features. But all examples I can find use a simple notation:

```ts
type Example1 = Dog extends Animal ? number : string;
//   ^?
```

While the example in this repo includes the type information in the source:

```ts
  const { say } = createQueryService({ service: ElizaService });
  //      ^? (property) say: UnaryHooks<SayRequest, SayResponse>
```

Do I need to type this out if I want to contribute? Do I _have_ to install vscode to make sure it's valid?

I tried pasting the extended annotation in the TypeScript [Twoslash demo](https://www.typescriptlang.org/dev/twoslash/) and in the [playground](https://www.typescriptlang.org/play?#code/AQYw9gdgzgLsDewoEMCewC+wC8CWoC5gBGAJgGYMBuAKGAHp7hmWA9AfmAAoAHAJzA8ApnxioAlEjREIAVwC2AIxE1VASwgwRAM2QghwAIIQ185ABsEdYObUA3IV3FE7YNQBNaGGhq19d+sAAImAA5sBCAB5aEO5QRiZmlvDWAO5gYNpOLm6eNN40YsLAAKKRyPI85kLEOMFhEdFCsfHGphbAnHJKIsBEsHwaobSMbOyqRQZlFVVCpHUASkKhZTyNMXEJ7ZZdCsp8fUgwgxDDNKPMHEA), but the extended annotation didn't parse, and I didn't see either picked up in CodeSandbox either (I suppose this needs some extra setup).

This leads me to believe that "orta.vscode-twoslash-queries" is a vscode extension that supports what's actually a _superset_ of Twoslash queries because it inserts the type information right in the source code, making it incompatible with the playground. 

Please correct me if I'm missing something, but from what I see, this is closely tied to vscode and the extension, which is what prompted the code smell discussion.

Don't get me wrong, I'd love visual type hints in examples, docs, and playgrounds. I'm wondering how to get there, and would like to discuss trade-offs (because there clearly are some) and options.

Reply:
@timostamm the editor integration independent of the documentation parser.  both use twoslash.  The TypeScript documentation uses [plain React](https://github.com/microsoft/TypeScript-Website/blob/v2/packages/typescriptlang-org/src/pages/dev/twoslash.tsx) along with [a shiki plugin](https://shikijs.github.io/twoslash/) which supports also [Docusaurus](https://github.com/shikijs/twoslash/tree/main/packages/docusaurus-preset-shiki-twoslash), [11ty](https://github.com/shikijs/twoslash/tree/main/packages/eleventy-plugin-shiki-twoslash), [Gatsby](https://github.com/shikijs/twoslash/tree/main/packages/gatsby-remark-shiki-twoslash), [Hexo](https://github.com/shikijs/twoslash/tree/main/packages/hexo-shiki-twoslash), [VuePress](https://github.com/shikijs/twoslash/tree/main/packages/vuepress-plugin-shiki-twoslash), [Markdown-It](https://github.com/shikijs/twoslash/tree/main/packages/markdown-it-shiki-twoslash), and [Remark](https://github.com/shikijs/twoslash/tree/main/packages/remark-shiki-twoslash), [and](https://github.com/shikijs/twoslash/tree/main/packages/shiki-twoslash) [more](https://github.com/shikijs/twoslash/tree/main/packages/twoslash-cli).

The editor integration uses [`inlay hints`](https://code.visualstudio.com/docs/typescript/typescript-editing#_inlay-hints) which allows text after the `^?` because inlay hints are inline.  The docs parser is not inline, so if you remove that part it works fine:

<img src="https://user-images.githubusercontent.com/15232461/212967042-2329ef66-42bd-4b0d-8ba3-f6cfd55f8db8.png" width="75%" />

Now that you're familiar with it, I bet you'll notice it other docs.  For example, [tRPC docs do this same thing in their docs](https://trpc.io/docs/quickstart#querying--mutating) ([source](https://github.com/trpc/trpc/blob/main/www/docs/main/quickstart.mdx)).

re: _"Do I need to type this out if I want to contribute?"_ no.
re: _"Do I have to install vscode to make sure it's valid?"_ no.  We should remove everything after the `^?`.  It was only there for me to double check before I had tests and not intended to be final.

Reply:
The arguments for and against Twoslash side, our general OSS principles do not think that checking in IDE files is a great thing to do. From many OSS users' perspective (including ours), it comes across as a code smell - we don't imagine there will be a ton of external contribution, and on balance it feels like a better move to remove the directory.

Reply:
Speaking as a very light OSS contributor, I have appreciated dedicated .vscode config that let's my editor start out already configed. Specifically things like pointing vscode's typescript version to the one locally installed in node_modules, things like that. Perhaps this is just a FE perspective but I've never considered it a code smell as long as it doesn't refer to local files outside of the project.